impl/src/main/java/com/okta/sdk/impl/oauth2/AccessTokenRetrieverServiceImpl.java

```diff
@@ -0,0 +1,184 @@
+
+    private ClientConfiguration clientConfiguration;
+
+    public AccessTokenRetrieverServiceImpl(ClientConfiguration clientConfiguration) {
+        this.clientConfiguration = clientConfiguration;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public OAuth2AccessToken getOAuth2AccessToken() throws IOException, InvalidKeyException {
+        log.debug("Getting OAuth2 access token for client id {} from {}",
+            clientConfiguration.getClientId(), clientConfiguration.getBaseUrl());
+
+        String signedJwt = createSignedJWT(clientConfiguration);
+        String scope = String.join(" ", clientConfiguration.getScopes());
+
+        QueryString queryString = new QueryString();
+        queryString.put("grant_type", "client_credentials");
+        queryString.put("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer");
+        queryString.put("client_assertion", signedJwt);
+        queryString.put("scope", scope);
+
+        HttpHeaders httpHeaders = new HttpHeaders();
+        httpHeaders.add("Accept", "application/json");
+        httpHeaders.add("Content-Type", "application/x-www-form-urlencoded");
+
+        Request accessTokenRequest = new DefaultRequest(
+            HttpMethod.POST,
+            clientConfiguration.getBaseUrl() + "/oauth2/v1/token",
+            queryString,
+            httpHeaders,
+            new ByteArrayInputStream("".getBytes("UTF-8")),
+            0);
+
+        RequestExecutor requestExecutor = new OkHttpRequestExecutorFactory().create(clientConfiguration);
+
+        Response accessTokenResponse = requestExecutor.executeRequest(accessTokenRequest);
+
+        if (accessTokenResponse != null && accessTokenResponse.getBody() != null) {
+            Reader reader = new InputStreamReader(accessTokenResponse.getBody(), StandardCharsets.UTF_8);
+
+            if (accessTokenResponse.getHttpStatus() == HttpURLConnection.HTTP_OK) {
+                log.debug("OAuth2 access token request was successful");
+                return objectMapper.readValue(reader, OAuth2AccessToken.class);
+            } else {
+                log.debug("OAuth2 access token request failed with HTTP {} error", accessTokenResponse.getHttpStatus());
+
+                OAuth2ErrorResponse errorResponse = objectMapper.readValue(reader, OAuth2ErrorResponse.class);
+
+                if (!Strings.isEmpty(errorResponse.getError()) ||
+                    !Strings.isEmpty(errorResponse.getErrorDescription())) {
+                    throw new HttpException("Received HTTP " + accessTokenResponse.getHttpStatus() +
+                        " response from Authorization Server. error - " + errorResponse.getError() +
+                        ", error description - " + errorResponse.getErrorDescription());
+                } else {
+                    throw new HttpException("Received HTTP " + accessTokenResponse.getHttpStatus() +
+                        " response from Authorization Server");
```

Comment:
Add something about when requesting an OAuth token

Reply:
I'm logging it already.

Reply:
yes, but its a different level and audience.
The application using this SDK is likely only logging at level `WARN`, so when they catch the exception they wouldn't have that context. so this exception should contain that information so they can understand the issue.

Reply:
If I understood correctly, you recommend logging exceptions like above at ERROR level and successful OAUTH token logging at INFO logger level, right?

Reply:
Nope, sorry I probably wasn't very clear.

The current debug logging is good, it just doesn't help the customer as much.

I'm suggesting adding a more detailed message to the exception, so the caller (the end user's application) has those details. Possibly throwing something like an `InvalidAuthenticationException` which can capture the status code into a typed object. And they can log the exception if/when they handle it.

Does that help?

Reply:
Sure, this helps!